msgraph-developer-proxy-plugins/RequestLogs/ExecutionSummaryPlugin.cs

```diff
@@ -0,0 +1,390 @@
+  private readonly Option<string?> _filePath;
+  private readonly Option<SummaryGroupBy?> _groupBy;
+  private const string _requestsInterceptedMessage = "Requests intercepted";
+  private const string _requestsPassedThroughMessage = "Requests passed through";
+
+  public ExecutionSummaryPlugin()
+  {
+    _filePath = new Option<string?>("--summary-file-path", "Path to the file where the summary should be saved. If not specified, the summary will be printed to the console. Path can be absolute or relative to the current working directory.");
+    _filePath.ArgumentHelpName = "summary-file-path";
+    _filePath.AddValidator(input =>
+    {
+      var outputFilePath = input.Tokens.First().Value;
+      if (string.IsNullOrEmpty(outputFilePath))
+      {
+        return;
+      }
+
+      var outputDir = Path.GetFullPath(Path.GetDirectoryName(outputFilePath));
+      if (!Directory.Exists(outputDir))
+      {
+        input.ErrorMessage = $"The directory {outputDir} does not exist.";
+      }
+    });
+
+    _groupBy = new Option<SummaryGroupBy?>("--summary-group-by", "Specifies how the information should be grouped in the summary. Available options: `url` (default), `messageType`.");
+    _groupBy.ArgumentHelpName = "summary-group-by";
+    _groupBy.AddValidator(input =>
+    {
+      if (!Enum.TryParse<SummaryGroupBy>(input.Tokens.First().Value, true, out var groupBy))
+      {
+        input.ErrorMessage = $"{input.Tokens.First().Value} is not a valid option to group by. Allowed values are: {string.Join(", ", Enum.GetNames(typeof(SummaryGroupBy)))}";
+      }
+    });
+  }
+
+  public override void Register(IPluginEvents pluginEvents,
+                          IProxyContext context,
+                          ISet<Regex> urlsToWatch,
+                          IConfigurationSection? configSection = null)
+  {
+    base.Register(pluginEvents, context, urlsToWatch, configSection);
+
+    configSection?.Bind(_configuration);
+
+    pluginEvents.Init += OnInit;
+    pluginEvents.OptionsLoaded += OnOptionsLoaded;
+    pluginEvents.AfterRecordingStop += AfterRecordingStop;
+  }
+
+  private void OnInit(object? sender, InitArgs e)
+  {
+    e.RootCommand.AddOption(_filePath);
+    e.RootCommand.AddOption(_groupBy);
+  }
+
+  private void OnOptionsLoaded(object? sender, OptionsLoadedArgs e)
+  {
+    InvocationContext context = e.Context;
+
+    var filePath = context.ParseResult.GetValueForOption(_filePath);
+    if (filePath is not null)
+    {
+      _configuration.FilePath = filePath;
+    }
+
+    var groupBy = context.ParseResult.GetValueForOption(_groupBy);
+    if (groupBy is not null)
+    {
+      _configuration.GroupBy = groupBy.Value;
+    }
+  }
+
+  private void AfterRecordingStop(object? sender, RecordingArgs e)
+  {
+    if (!e.RequestLogs.Any())
+    {
+      return;
+    }
+
+    var report = _configuration.GroupBy switch
+    {
+      SummaryGroupBy.Url => GetGroupedByUrlReport(e.RequestLogs),
```

Comment:
Love these new expression body pattern switches, so terse and expressive

Reply:
Pretty cool, huh? Brought to you by Copilot 🚀